lib/interfaces/include/BMSDriverGroup.h

```diff
@@ -0,0 +1,332 @@
+     * NOTE: Conversions are different depending on which we are reading.
+     * @pre in order to actually "read" anything, we need to call wakeup() and send data over SPI
+     * @post store all temperature data into the board_temperatures container
+     * AND record the maximum value and locations
+     */
+    void read_GPIOs();
+
+    /* -------------------- WRITING DATA FUNCTIONS -------------------- */
+
+    /**
+     * Writes the device configuration
+     * @pre needs access to undervoltage, overvoltage, configuration MACROS, and discharge data
+     * @post sends packaged data over SPI
+     */
+    void write_configuration(uint8_t dcto_mode);
+
+    /**
+     * Writes command to start cell voltage ADC converion
+     * @post packaged data transferred over SPI, need to delay before we can read
+     */
+    void start_cell_voltage_ADC_conversion();
+
+    /**
+     * Writes command to start GPIO ADC conversion
+     * @post packaged data transfered over SPI, need to delay before we can read
+     * This means that we have to call a delay before continuing.
+     */
+    void start_GPIO_ADC_conversion();
+
+    /* -------------------- PRINT DATA FUNCTIONS -------------------- */
+
+    /**
+     * Prints all the cell voltage information
+     * Every cell voltage, the minimum, maximum, and locations of those
+     */
+    void print_voltage_data();
+
+    /**
+     * @post resets the max, min data holders to outside bound
+     */
+    void reset_voltage_data();
+
+    /**
+     * Prints all the GPIO information
+     * Every temperature, humidity, the maximum temperature, and locations of those
+     */
+    void print_GPIO_data();
+
+    /**
+     * @post resets the max, min data holders to outside bound
+     */
+    void reset_GPIO_data();
+
+    /* -------------------- SETTER FUNCTIONS -------------------- */
+
+    /**
+     * Sets chip select for this IC
+     * @param integer input chip select
+     */
+    void set_chip_select(int ic_index, int cs) { this->chip_select = cs; }
+
+    /**
+     * ADDRESSES: Addresses only pertain to LTC6811-2s which are encoded in hardware through 4 bits
+     * These addresses need to be coded into the firmware so that address commands are properly sent
+     * This function is a setter for the address
+     * @param integer input address
+     */
+    void set_address(int *addr) { this->address = addr; }
+
+    /* -------------------- GETTER FUNCTIONS -------------------- */
+
+    /**
+     * Generates a Packet Error Code
+     * @pre PEC table is already generated –> pec15Table has data generated in it
+     * @post returns an unsigned 16 bit integer used to check data to and from the device
+     * @param data data we are writing PEC for, could be Command Code or Buffer data
+     * @param length length of data
+     * @return unsigned 16 bit PEC, array of uint8_t of length 2
+     */
+    uint8_t *calculate_specific_PEC(uint8_t *data, int length);
```

Comment:
this should return a `std::array<uint8_t, 2>`, and the data in can also be passed in via a std::array like this:

```c++
template <size_t data_len>
std::array<uint8_t, 2> calculate_specific_PEC(const std::array<uint8_t, data_len> &data)
```